### PR TITLE
`binary_cross_entropy`: double backwart wrt target

### PR DESCRIPTION
**Pull request by nikitaved**
As per title. An effort to make `binary_cross_entropy` all around differentiable.

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -1578,6 +1578,42 @@ Tensor binary_cross_entropy_target_backward(
...
+    if (!isTensorSubclassLike(weight.value())) {
+      res.mul_(weight.value());
+    } else {
+      res = res * weight.value();
+    }
+  }
+
+  auto denom = (1 - self).mul_(self);
+  {
+    at::NoGradGuard guard;
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:1601]: Compliance wrt to other kernels, which clamp at `eps` in an autograd-agnostic way.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -1578,6 +1578,42 @@ Tensor binary_cross_entropy_target_backward(
...
+    // Default eps in binary_cross_entropy for ALL dtypes
+    // TODO: probably change this to a dtype-dependent value
+    double eps = 1e-12;
+    denom.clamp_min_(eps);
+  }
+
+  res = res.div_(denom);
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:1608]: Could you please explain why this line may cause any issues? Neither `res` nor `denom` modify `self` in-place... Or am I missing something?

[the remark this answers is not in the document]
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:1608]: Oh, I see! That makes sense! Sorry, missed out on that...

[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.cpp:1608]: Yeah conditioning on `denom` might be more precise actually. Conditioning on `self` would need to assume that `(1 - self).mul_(self)` preserved the subclass.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.cpp:1601]: Oh gotcha, also from discussion with @albanD offline, it seems like it is useful for avoiding setting gradients to zero for values that are clamped. Not sure the clamp is entirely for division by zero in this case though because negative values would also be clamped to eps?

[the remark this answers is not in the document]
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:1601]: There should be no negative values, inputs for binary cross entropy are values between 0 and 1.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:1608]: Thank you, will add this check there too then.

[the remark this answers is not in the document]
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:1608]: @soulitzer , can we assume that `1 - self` is safe?

File: test/test_nn.py [hunk 3]
```diff
@@ -9940,22 +9940,6 @@ def func(x):
             # just run a single backward, as gradcheck/gradgradcheck is expensive here
             output.sum().backward()
 
-    def test_binary_cross_entropy_grads(self):
-        import torch.nn.functional as F
```
Comment (nikitaved) [inline test/test_nn.py:9944]: No need: we have OpInfos now.

**Comment by nikitaved** [comment]
@soulitzer , could you please have a look again?

**Review by soulitzer — APPROVED** [review]
Sorry for the delay, looks good!
